lib/commons/dom/is-visible.js

```diff
@@ -13,14 +13,187 @@ const clipRegex = /rect\s*\(([0-9]+)px,?\s*([0-9]+)px,?\s*([0-9]+)px,?\s*([0-9]+
 const clipPathRegex = /(\w+)\((\d+)/;
 
 /**
- * Determines if an element is hidden with a clip or clip-path technique
- * @method isClipped
+ * Determine whether an element is visible
+ * @method isVisible
  * @memberof axe.commons.dom
- * @private
- * @param  {CSSStyleDeclaration} style Computed style
- * @return {Boolean}
+ * @instance
+ * @param {HTMLElement} el The HTMLElement
+ * @param {Boolean} screenReader When provided, will evaluate visibility from the perspective of a screen reader
+ * @param {Boolean} recursed
+ * @return {Boolean} The element's visibilty status
+ */
+export default function isVisible(el, screenReader, recursed) {
+  if (!el) {
+    throw new TypeError(
+      'Cannot determine if element is visible for non-DOM nodes'
+    );
+  }
+
+  const vNode = el instanceof AbstractVirtualNode ? el : getNodeFromTree(el);
+  el = vNode ? vNode.actualNode : el;
+
+  const cacheName = '_isVisible' + (screenReader ? 'ScreenReader' : '');
+  const { DOCUMENT_NODE, DOCUMENT_FRAGMENT_NODE } = window.Node ?? {};
+  const nodeType = vNode ? vNode.props.nodeType : el.nodeType;
+  const nodeName = vNode ? vNode.props.nodeName : el.nodeName.toLowerCase();
```

Comment:
Why wouldn't there be a virtual node? Is this because you're recursing on document fragments? Seems like that can be avoided.

Reply:
Added a comment. TL;DR is that we recurse up the DOM tree and can find parent elements not in the virtual node tree, but we still need to know the visible state of all elements.